tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,195 @@
+            // argument is optional. If none is specified, the default machine parameters
+            // for a generic CPU architecture are going to be used by the auto-scheduler.
+
+            // Let's use some arbitrary but plausible values for the machine parameters.
+            MachineParams machine_params(32, 16 * 1024 * 1024, 40);
+            // The arguments to MachineParams are the maximum level of parallelism
+            // available, the size of the last-level cache (in KB), and the ratio
+            // between memory cost and arithmetic cost at the last level case, of
+            // the target architecture, in that order.
+
+            // Note that when using the auto-scheduler, no schedule should have
+            // been applied to the pipeline; otherwise, the auto-scheduler will
+            // throw an error. The current auto-scheduler does not work with
+            // partially-scheduled pipeline.
+            //
+            // Calling auto_schedule_outputs() will apply the generated schedule
+            // automatically to members of the pipeline in addition to returning
+            // a string representation of the schedule.
+            std::string schedule = auto_schedule_outputs(machine_params);
+            std::cout << "\nSchedule:\n\n" << schedule << "\n";
+
+            // The generated schedule that is dumped to std::cout is an actual
+            // Halide C++ source, which is readily copy-pastable back into
+            // this very same source file with little modifications. Programmers
+            // can use this as a starting schedule and iteratively improve the
+            // schedule. Note that the current auto-scheduler is only able to
+            // generate CPU schedule and only does tiling, simple vectorization
+            // and parallelization. It doesn't deal with line buffering, storage
+            // reordering, or factoring a reduction.
+        } else {
+            // This is where you would declare the schedule you have or
+            // paste the schedule generated by the auto-scheduler.
+
+            // This auto-scheduler will return the following schedule for the
+            // estimates and machine parameters declared above when run on
+            // this pipeline:
+            //
+            // Var x_vi("x_vi");
+            // Var x_vo("x_vo");
+            //
+            // Func f0 = pipeline.get_func(3);
+            // Func output1 = pipeline.get_func(4);
+            // Func output2 = pipeline.get_func(5);
+            //
+            // {
+            //     Var x = f0.args()[0];
+            //     Var y = f0.args()[1];
+            //     Var z = f0.args()[2];
+            //     Var n = f0.args()[3];
+            //     RVar r$x(f0.update(0).get_schedule().rvars()[0].var);
+            //     RVar r$y(f0.update(0).get_schedule().rvars()[1].var);
+            //     RVar r$z(f0.update(0).get_schedule().rvars()[2].var);
```

Comment:
Not specific to this tutorial PR, but it seems these RVars are never used in the generated schedule. Is it worth adding extra logic to prune dead code and clean up the generated schedule string?

Reply:
Fixed in #2408